models/interactions.sql

```diff
@@ -0,0 +1,195 @@
+    on mp.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_youtube_youtube_1') }} as y
+    on  y.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_media_element_1') }} as me
+    on me.root_id = e.event_id
+
+    left join {{ source('atomic', 'org_whatwg_video_element_1') }} as ve
+    on ve.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_video_1') }} as r
+    on r.root_id = e.event_id
+
+    left join {{ source('atomic', 'com_roku_device_info_1') }} as rd
+	  on  rd.root_id = e.event_id
+
+)
+
+, interaction_ends as (
+
+	select
+		page_view_id,
+		media_id,
+		event_id,
+		event_type,
+		start_tstamp,
+		media_start_pos,
+		lead(start_tstamp, 1) over(partition by page_view_id, media_id order by start_tstamp) end_tstamp,
+		lead(media_start_pos, 1) over(partition by page_view_id, media_id order by start_tstamp) media_end_pos	
+		
+	from prep	
+
+)
+
+, find_nulls_round_1 as (
+
+	select 
+		event_id,
+    start_tstamp,
+    event_type,
+    case when event_type in ('pause', 'paused', 'ended') then coalesce(end_tstamp, start_tstamp) else end_tstamp end end_tstamp,
+    media_start_pos,
+		case when event_type in ('pause', 'paused', 'ended') then coalesce(media_end_pos, media_start_pos) else media_end_pos end media_end_pos
+	
+	from interaction_ends
+
+)
+
+-- for now default it to start_tstamp, same as above but it could be the page_view end_tstamp
+, find_nulls_round_2 as (
+
+	select
+		f.event_id,
+    f.event_type,
+    f.start_tstamp,
+		case when f.end_tstamp is null then p.start_tstamp else f.end_tstamp end end_tstamp,
+    f.media_start_pos,
+		case when f.media_end_pos is null then p.media_start_pos else f.media_end_pos end as media_end_pos 
+	
+	from find_nulls_round_1 f
+
+	inner join prep p
+	on p.event_id = f.event_id
+
+)	 
+
+, time_periods as (
+
+    select
+		  r.event_id,
+      r.event_type,
+      r.start_tstamp,
+      r.end_tstamp,
+      r.media_start_pos,
+      r.media_end_pos,
+      datediff(second,r.start_tstamp, r.end_tstamp) as elapsed_time_sec,
+	    case when r.media_end_pos - r.media_start_pos < 1 then 0 else round(r.media_end_pos - r.media_start_pos) / p.playback_rate end as play_time_sec -- this is to remove negative play_time_secs which are anyway the result of seeks which we want to exclude
+
+    from find_nulls_round_2 r
+
+    left join prep p
+	  on p.event_id = r.event_id
+
+)
+
+, corrected_play_time as (
+
+   select
+    event_id,
+    media_end_pos,
+    end_tstamp,
+    case 
+      when event_type in ('pause', 'paused', 'ended', 'seek', 'seeked') then 0
+      when (play_time_sec - elapsed_time_sec) > 1 then elapsed_time_sec -- to correct cases when seek disrupts the event order based play_time_end_sec calculation
+      else play_time_sec
+    end as play_time_sec_amended,
+    elapsed_time_sec,
+    play_time_sec
+
+  from time_periods
+
+)
+
+, corrected_duration as (
+
+  select
+    media_id,
+    max(duration) as duration
```

Comment:
I have noticed a few edge cases where the event came without a duration context for some unknown reason hence the  calculation here. Also, in cases it is a fraction not a whole number hence the earlier rounding on top of this. Thought it is safer to make it coherent, especially that this will be used as a base for calculations later on.